Add blog-post about downtream workflow (#404)
packit/packit.dev: content/posts/downstream-automation.md

---

> title: "Downstream automation is here"

Reviewer: Thank you for writing this down!

Is this meant to be a "heads-up" to our users or a guide that we'll update and refer to in future?

Author: I would say `heads-up`, but now I am not sure what's better...;)

It's true that documentation of separate jobs is not enough to see the whole workflow.

Reviewer: My point is: we need a proper guide, very user-friendly, to describe in detail how to bring upstream releases into Fedora (using Packit) :)

Author: I see it can act as one. So, we can either convert this to that guide or make a slightly changed version of this that will be preserved/updated and linked in the future. I don't like duplicates, but in this case, I would rather create a separate guide.

Reviewer: yup, sounds good, let's keep it and once we start working on the guide, we can get inspired here and even change/update it (for reference packit/packit.dev#351)

Let's take a look at how this works on an example of [OGR](https://github.com/packit/ogr), the Python library we develop:

## Upstream

The process of releasing a new version starts in the upstream repository.
Here, we can see an upstream release:

TODO: upstream release image

## Propose downstream

As the first step on our way to Fedora users, we need to get the new upstream release to the Fedora dist-git.
This is what we call `propose-downstream` job.
Here is a snippet from the config file of OGR:

```yaml
downstream_package_name: python-ogr
copy_upstream_release_description: true

jobs:
  - job: propose_downstream
    trigger: release
    metadata:
      dist_git_branches:
        - fedora-all
        - epel-8
```

And how does this work?
Packit gets the information about the newly created release from GitHub (via webhook).
load the config from the release commit and if there is a `propose-downstream` job
defined, the workflow begins.
As a first step, the archive is saved to lookaside cache
and after that, Packit updates the dist-git content (mainly `sources` file and spec-file)
via pull-requests for the specified branches. (Direct push is possible by setting a

> [`create_pr` option](https://packit.dev/docs/configuration/#create_pr) to `false`.)

Reviewer: Could we split this paragraph into two sections?
1. What work is happening behind the scenes (clone & checkout, update spec file & %changelog, push, sources, create PRs)
2. The implementation details - how it works

Author: Good point. PTAL

Reviewer: Thanks Franto! Nicely done

Finally, it's here. Now, you can do the whole Fedora release with the help of Packit.
Let's take a look at how this works on an example of [OGR](https://github.com/packit/ogr), the Python library we develop:

## Upstream

The process of releasing a new version starts in the upstream repository.
Here, we can see an upstream release:

TODO: upstream release image

## Propose downstream

As the first step on our way to Fedora users, we need to get the new upstream release to the Fedora dist-git.
This is what we call `propose-downstream` job.
Here is a snippet from the config file of OGR:

```yaml
downstream_package_name: python-ogr
copy_upstream_release_description: true

jobs:
  - job: propose_downstream
    trigger: release
    metadata:
      dist_git_branches:
        - fedora-all
        - epel-8
```

How does the `propose-downstream` work?
As a first step, the archive is saved to lookaside cache
and after that, Packit updates the dist-git content (mainly `sources` file and spec-file)
via pull-requests for the specified branches. (Direct push is possible by setting a
[`create_pr` option](https://packit.dev/docs/configuration/#create_pr) to `false`.)

> As you can see in the config above, if you use `copy_upstream_release_description: true`,

Reviewer 2: ```suggestion
If you use [`copy_upstream_release_description: true`](https://packit.dev/docs/configuration/#copy_upstream_release_description), as in the config above,
```

the changelog entry will use the GitHub release description field.
(Just make sure the formatting is compatible with spec-file.
E.g. use `-` instead of `*` for lists not to create multiple changelog entries.)

And how it is triggered?
Packit gets the information about the newly created release from GitHub (via webhook),
load the config from the release commit and if there is a `propose-downstream` job
defined, the workflow begins.

Here is an example pull-request created by Packit:

TODO: example dist-git PR image

Now, it's on downstream CI systems and maintainer to check the changes and merge
the pull-request.

## Koji

> If Packit sees a new commit in the configured branch, it submits a new build in Koji

Reviewer 2: just to emphasize that here we are downstream

```suggestion
If Packit sees a new commit in the configured dist-git branch, it submits a new build in Koji
```

Here is a job definition for the package we use as an example:

```yaml
jobs:
  - job: koji_build
    trigger: commit
    metadata:
      dist_git_branches:
        - fedora-all
        - epel-8
```

There is no UI provided by Packit for the job,
but it is visible across Fedora systems (like you can see in the following image)
like a manually created Koji build and you can utilise
[Fedora Notifications](https://apps.fedoraproject.org/notifications/about)
to get informed about the builds.

TODO: example of Koji status in the dist-git image

## Bodhi

Once Packit is informed (via fedora-messaging bus) about the successful Koji build,
it creates a new update in Bodhi for you.

Here is a job definition:

```yaml
jobs:
  - job: bodhi_update
    trigger: commit
    metadata:
      dist_git_branches:
        - fedora-stable # rawhide updates are created automatically
        - epel-8
```

The packit config is loaded from the commit the build is triggered from.

Here is an example of the resulting Bodhi update

TODO: Bodhi update image

And that's all. The rest is on the users and maintainers to give the update enough Karma

> so it reach the users.

Reviewer 2: ```suggestion
so the update gets to the users.
```

---
title: "Downstream automation is here"
date: 2022-03-02
weight: 68
---

# Downstream automation is here

Finally, it's here. Now, you can do the whole Fedora release with the help of Packit.

> Let's take a look at how this works on an example of [OGR](https://github.com/packit/ogr), the Python library we develop:

Reviewer 3: ```suggestion
Let's take a look at how it works on an example of [OGR](https://github.com/packit/ogr), the Python library we develop.
```

## Upstream

The process of releasing a new version starts in the upstream repository.
Here, we can see an upstream release:

![Upstream release](/images/downstream_workflow_upstream_release.png)

## Propose downstream

As the first step on our way to Fedora users, we need to get the new upstream release to the Fedora dist-git.
This is what we call `propose-downstream` job.
Here is a snippet from the config file of OGR:

```yaml
downstream_package_name: python-ogr
copy_upstream_release_description: true

jobs:
  - job: propose_downstream
    trigger: release
    metadata:
      dist_git_branches:
        - fedora-all
        - epel-8
```

How does the `propose-downstream` work?
As a first step, the archive is saved to lookaside cache
and after that, Packit updates the dist-git content (mainly `sources` file and spec-file)
via pull-requests for the specified branches. (Direct push is possible by setting a

> [`create_pr` option](https://packit.dev/docs/configuration/#create_pr) to `false`.)

Reviewer 3: only CLI

If you use [`copy_upstream_release_description: true`](https://packit.dev/docs/configuration/#copy_upstream_release_description),
as in the config above,
the changelog entry will use the GitHub release description field.
(Just make sure the formatting is compatible with spec-file.

> E.g. use `-` instead of `*` for lists not to create multiple changelog entries.)

Reviewer 3: ```suggestion
E.g. use `-` instead of `*` for lists to not create multiple changelog entries.)
```

> And how it is triggered?

Reviewer 3: ```suggestion
And how is it triggered?
```

Packit gets the information about the newly created release from GitHub (via webhook),

> load the config from the release commit and if there is a `propose-downstream` job

Reviewer 3: ```suggestion
loads the config from the release commit and if there is a `propose-downstream` job
```
